### PR TITLE
Add leeway for time claims in token validation

### PR DESCRIPTION
**Pull request by thiva-k**
This pull request introduces configurable leeway (in seconds) for time-based JWT validations across the authentication and OAuth subsystems, allowing the system to tolerate small clock skews between servers. The leeway value is now configurable via the main config and is respected in all relevant JWT claim checks (`exp`, `nbf`, and `iat`). Comprehensive unit tests have been added to ensure correct behavior for tokens near expiration, not-yet-valid, or issued in the future, both within and beyond the leeway window.

**Leeway configuration and usage:**

* Added a new `leeway` parameter to the JWT configuration (`JWTConfig`) and the default config file (`default.json`), enabling system-wide clock skew tolerance for JWT validation. [[1]](diffhunk://#diff-beec713bb8f44ef8633de8d745de6dcd555593b5cd0758d7788470fee3b2d974L75-R76) [[2]](diffhunk://#diff-70ef91d44215280f2234659532a25f97986a90f9f6e8e798167394efdfe0bacbR112)
* Updated JWT claim validation logic in the Google OIDC authentication service, OAuth token validator, and core JWT service to use the configured leeway when checking `exp`, `nbf`, and `iat` claims. [[1]](diffhunk://#diff-a5b044241b22c362b38508dd09ad0b565ec9b55d5e87e3871a4283fb03b37dc0R162-R184) [[2]](diffhunk://#diff-ef8e3cff68ea3a4ac66f0ee805abcd05c771536d7c66506883aff62d976e76cfR195-R209) [[3]](diffhunk://#diff-a6adf3395a1977c374db36e94ae9a881988468a4e6610edf2e947a982751dd2dR446-R453) [[4]](diffhunk://#diff-a6adf3395a1977c374db36e94ae9a881988468a4e6610edf2e947a982751dd2dL460-R463)

**Testing and reliability:**

* Added and updated test setup to initialize the leeway configuration for all relevant test suites. [[1]](diffhunk://#diff-07fa6b7d966fbed03882ff4fe8f143b9e4781250a0eb242c52d6bb9e1ae122dfR64-R71) [[2]](diffhunk://#diff-e54fae60295c6be9e1651f6e17000855cee2420b4f41279f560721325cb22d1aR61) [[3]](diffhunk://#diff-17d57f04a8044070e7b47bdfca7200f357eacd8797bc7a5d357b67c5688537a1R139)
* Implemented extensive new unit tests for the Google OIDC service, OAuth token validator, and JWT service to verify correct leeway handling for tokens expiring, not yet valid, or issued in the future, including edge cases for zero and non-zero leeway. [[1]](diffhunk://#diff-07fa6b7d966fbed03882ff4fe8f143b9e4781250a0eb242c52d6bb9e1ae122dfR803-R941) [[2]](diffhunk://#diff-e54fae60295c6be9e1651f6e17000855cee2420b4f41279f560721325cb22d1aR1374-R1500) [[3]](diffhunk://#diff-17d57f04a8044070e7b47bdfca7200f357eacd8797bc7a5d357b67c5688537a1R2314-R2454)

**Codebase consistency:**

* Ensured all relevant imports and initializations are updated to support the new leeway configuration. [[1]](diffhunk://#diff-a5b044241b22c362b38508dd09ad0b565ec9b55d5e87e3871a4283fb03b37dc0R30) [[2]](diffhunk://#diff-07fa6b7d966fbed03882ff4fe8f143b9e4781250a0eb242c52d6bb9e1ae122dfR31)

These changes make JWT validation more robust in distributed environments by preventing unnecessary authentication failures due to minor time discrepancies between systems.

Related issue:
- https://github.com/asgardeo/thunder/issues/1183

### CODE REVIEW

File: backend/internal/system/jwt/service.go [hunk 1]
```diff
@@ -443,11 +443,14 @@ func (js *jwtService) verifyJWTClaims(jwtToken string, expectedAud, expectedIss
...
 
+	// Get leeway from config to account for clock skew
+	leeway := config.GetThunderRuntime().Config.JWT.Leeway
+
 	// Validate standard claims (exp, nbf, aud, iss)
 	now := time.Now().Unix()
 
 	if exp, ok := payload["exp"].(float64); ok {
-		if now > int64(exp) {
+		if now > int64(exp)+leeway {
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/system/jwt/service.go]: Fixed

File: backend/internal/system/config/config.go [hunk 2]
```diff
@@ -109,6 +109,7 @@ type JWTConfig struct {
 	ValidityPeriod int64  `yaml:"validity_period" json:"validity_period"`
 	Audience       string `yaml:"audience" json:"audience"`
 	PreferredKeyID string `yaml:"preferred_key_id" json:"preferred_key_id"`
+	Leeway         int64  `yaml:"leeway" json:"leeway"`
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/system/config/config.go:112]: Not needed

File: backend/internal/system/jwt/service_test.go [hunk 3]
```diff
@@ -2310,3 +2311,144 @@ func (suite *JWTServiceTestSuite) TestVerifyJWTSignatureWithPublicKeyAlgorithmDe
...
+				expiredTime := time.Now().Add(-31 * time.Second).Unix()
+				token := suite.createBasicJWT(aud, iss, expiredTime, time.Now().Add(-time.Hour).Unix())
+				return token, aud, iss
+			},
+			setupConfig: func() {
+				// Leeway of 30 seconds is already configured in SetupTest
+			},
+			expectError:   true,
+			expectedError: ErrorTokenExpired,
+		},
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/system/jwt/service_test.go:2401]: Irrelevant

File: backend/internal/system/config/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/system/config/config.go:112]: Irrelevant

File: backend/internal/system/jwt/service_test.go [hunk 4]
```diff
@@ -2310,3 +2311,144 @@ func (suite *JWTServiceTestSuite) TestVerifyJWTSignatureWithPublicKeyAlgorithmDe
...
+			expectError:   true,
+			expectedError: ErrorInvalidJWTFormat,
+		},
+		{
+			name: "TokenExpiredExactlyAtLeewayBoundary_ShouldFail",
+			setupFunc: func() (string, string, string) {
+				aud := testAudience
+				iss := testIssuer
+				// Token expired exactly 31 seconds ago (just beyond 30s leeway)
+				expiredTime := time.Now().Add(-31 * time.Second).Unix()
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/system/jwt/service_test.go:2392]: This is fine

File: backend/cmd/server/repository/resources/conf/default.json [hunk 5]
```diff
@@ -72,7 +72,8 @@
   "jwt": {
     "validity_period": 3600,
     "audience": "application",
-    "preferred_key_id": "default-key"
+    "preferred_key_id": "default-key",
+    "leeway": 30
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/cmd/server/repository/resources/conf/default.json:76]: Irrelevant comment
